### PR TITLE
[JUJU-422] Remove space.Subnets method

### PR DESCRIPTION
**Pull request by jack-w-shaw**
As stated [here](https://github.com/juju/juju/blob/207cf067e2ee4727fb072288cb4cd6af47729c2f/state/spaces.go#L70-L71), the method space.Subnets should be phased out.

Remove all remaining invocations of space.Subnets, and finally remove the method itself

This has also opened the possibility for some pretty large reductions in complexity and the removal of redundancies. A couple of shims and the like have been removed

## Checklist

 - [?] Requires a [pylibjuju](https://github.com/juju/python-libjuju) change
 - [x] Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR
 - [?] Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed
 - [x] Comments answer the question of why design decisions were made

## QA steps

```sh
make static-analysis
go test github.com/juju/juju/apiserver/common/networkingcommon
go test github.com/juju/juju/apiserver/facades/client/...
go test github.com/juju/juju/apiserver/testing/
go test github.com/juju/juju/rpc/params/
go test github.com/juju/juju/core/network/
go test github.com/juju/juju/state/ --test.timeout=1800s
```

### Also

Bootstrap a controller into aws

#### List spaces
```sh
$ juju spaces
Name   Space ID  Subnets       
alpha  0         172.31.0.0/20 
                 172.31.16.0/20
                 172.31.32.0/20
                 252.0.0.0/12  
                 252.16.0.0/12 
                 252.32.0.0/12 
```

#### List subnets
```sh
$ juju subnets
subnets:
  172.31.0.0/20:
    type: ipv4
    provider-id: subnet-0815d22b08e3d82d1
    provider-network-id: vpc-0d3a33e664f1a5715
    status: in-use
    space: alpha
    zones:
    - eu-west-2c
...
  252.32.0.0/12:
    type: ipv4
    provider-id: subnet-039d317932e61e450-INFAN-172-31-32-0-20
    provider-network-id: vpc-0d3a33e664f1a5715
    status: in-use
    space: alpha
    zones:
    - eu-west-2b
```

#### Add subnet

```sh
$ juju add-subnet 172.31.48.0/24 alpha
added subnet with CIDR "172.31.48.0/24" in space "alpha"

$ juju subnets
subnets:
  172.31.0.0/20:
    type: ipv4
    provider-id: subnet-0815d22b08e3d82d1
    provider-network-id: vpc-0d3a33e664f1a5715
    status: in-use
    space: alpha
    zones:
    - eu-west-2c
...
  172.31.48.0/24:
    type: ipv4
    provider-id: subnet-03d148ead6fa5625a
    provider-network-id: vpc-0d3a33e664f1a5715
    status: in-use
    space: alpha
    zones:
    - eu-west-2b
...
  252.32.0.0/12:
    type: ipv4
    provider-id: subnet-039d317932e61e450-INFAN-172-31-32-0-20
    provider-network-id: vpc-0d3a33e664f1a5715
    status: in-use
    space: alpha
    zones:
    - eu-west-2

$ juju spaces
Name   Space ID  Subnets       
alpha  0         172.31.0.0/20 
                 172.31.16.0/20
                 172.31.32.0/20
                 172.31.48.0/24
                 252.0.0.0/12  
                 252.16.0.0/12 
                 252.32.0.0/12 
```

#### Add space

```sh
$ juju add-space beta
added space "beta" with no subnets

$ juju spaces
Name   Space ID  Subnets       
alpha  0         172.31.0.0/20 
                 172.31.16.0/20
                 172.31.32.0/20
                 172.31.48.0/24
                 252.0.0.0/12  
                 252.16.0.0/12 
                 252.32.0.0/12 
beta   1                       
```

#### Move subnet

```sh
$ juju move-to-space beta 172.31.48.0/24
Subnet 172.31.48.0/24 moved from alpha to beta

$ juju spaces
Name   Space ID  Subnets       
alpha  0         172.31.0.0/20 
                 172.31.16.0/20
                 172.31.32.0/20
                 252.0.0.0/12  
                 252.16.0.0/12 
                 252.32.0.0/12 
beta   1         172.31.48.0/24
```

## Documentation changes

N/A

## Bug reference

N/A

### CODE REVIEW

File: apiserver/common/networkingcommon/types.go [hunk 1]
```diff
@@ -87,13 +87,28 @@ type BackingSpace interface {
...
-	// Subnets returns the subnets in the space
-	Subnets() ([]BackingSubnet, error)
+	// NetworkSpace maps the space into network.SpaceInfo
+	NetworkSpace() (BackingSpaceInfo, error)
 
 	// ProviderId returns the network ID of the provider
 	ProviderId() network.Id
 }
 
+type BackingSpaceInfo struct {
```
Comment (manadart) [inline apiserver/common/networkingcommon/types.go:97]: Don't need this; just have the backing return the `SpaceInfo`.

File: apiserver/facades/client/applicationoffers/conversions.go [hunk 2]
```diff
@@ -33,27 +30,3 @@ func paramsFromProviderSpaceInfo(info *environs.ProviderSpaceInfo) params.Remote
 	}
 	return result
 }
-
-// spaceInfoFromState converts a state.Space into the equivalent
-// network.SpaceInfo.
-func spaceInfoFromState(space Space) (*network.SpaceInfo, error) {
```
Comment (manadart) [inline apiserver/facades/client/applicationoffers/conversions.go:39]: Nice find.

**Review by manadart — CHANGES_REQUESTED** [review]
Looks good, we just don't want the additions to the common types/shims.

File: apiserver/common/networkingcommon/types.go (hunk 1 above)
Comment (jack-w-shaw) [reply apiserver/common/networkingcommon/types.go:97]: Are you sure? The `Subnets` field needs to be of type []BackingSubnetInfo, not []SubnetInfo right?

File: state/subnets.go [hunk 3]
```diff
@@ -43,8 +44,8 @@ type subnetDoc struct {
 }
 
 // Life returns whether the subnet is Alive, Dying or Dead.
-func (s *Subnet) Life() Life {
-	return s.doc.Life
+func (s *Subnet) Life() life.Value {
```
Comment (manadart) [inline state/subnets.go:47]: I'd be inclined to leave this so it is consistent with the other `state` types, such as `Machine` and 'Unit`.

File: state/subnets.go [hunk 4]
```diff
@@ -361,6 +362,7 @@ func (s *Subnet) networkSubnet() network.SubnetInfo {
 		FanInfo:           fanInfo,
 		IsPublic:          s.doc.IsPublic,
 		SpaceID:           s.spaceID,
+		Life:              s.Life(),
```
Comment (manadart) [inline state/subnets.go:365]: Then just take `s.Life.Value()` here.

File: state/machine_linklayerdevices.go [hunk 5]
```diff
@@ -709,7 +710,7 @@ func (m *Machine) newIPAddressDocFromArgs(args *LinkLayerDeviceAddress) (*ipAddr
 }
 
 func (m *Machine) verifySubnetAlive(subnet *Subnet) error {
-	if subnet.Life() != Alive {
+	if subnet.Life() != life.Alive {
```
Comment (manadart) [inline state/machine_linklayerdevices.go:713]: And this file need not change.

File: core/network/subnet.go [hunk 6]
```diff
@@ -9,6 +9,8 @@ import (
 	"sort"
 	"strings"
 
+	"github.com/juju/juju/core/life"
```
Comment (manadart) [inline core/network/subnet.go:12]: Wrong group.

File: core/network/import_test.go [hunk 7]
```diff
@@ -8,14 +8,19 @@ import (
 	gc "gopkg.in/check.v1"
 
 	coretesting "github.com/juju/juju/testing"
+
+	"github.com/juju/collections/set"
```
Comment (manadart) [inline core/network/import_test.go:12]: Wrong group.

**Review by manadart — CHANGES_REQUESTED** [review]
Almost there.

**Comment by jack-w-shaw** [comment]
!!build!!

**Review by manadart — APPROVED** [review]
Nice. I'd like to look at a follow-up that replaces `BackingSubnetInfo` with `SubnetInfo` from `core`.

**Comment by jack-w-shaw** [comment]
$merge$

**Comment by jack-w-shaw** [comment]
$$merge$$

**Comment by jack-w-shaw** [comment]
$$merge$$

**Comment by jack-w-shaw** [comment]
$$merge$$

**Comment by jack-w-shaw** [comment]
$$merge$$

**Pull request merged**